### PR TITLE
Server run thread safety fix [changelog skip]

### PR DESCRIPTION
**Pull request by wjordan**
### Description
This PR fixes a race-condition creating the `@check, @notify` pipe in `Server`, which caused various intermittent test failures on JRuby.

The issue can be triggered by a call to `#run` (which calls `#handle_servers` in a background-thread) followed immediately by `#stop` (which calls `#notify_safely`). This can result in two separate `@check, @notify` pipes getting created concurrently, causing the signal sent by `#stop` to not get picked up by the listen-loop in `#handle_servers`, and therefore fail to exit as expected. This PR adds a test `test_run_stop_thread_safety` demonstrating the race condition that consistently fails on JRuby without this fix (see [this test run](https://github.com/wjordan/puma/actions/runs/314657885)).

The fix moves pipe-creation into `Server#run` (before the background thread is spawned), making the behavior more consistent that the server will stop as long as the call to `#stop` comes after the call to `#run` returns.

This PR includes a related timing change in `Single`, moving the `Use Ctrl-C to stop` log message after the call to `Server#run`. This is because the `wait_for_server_to_boot` integration-test helper uses this log message to determine when the server is 'booted', and some integration tests (e.g., `test_int_refuse`) send an `INT` signal (which calls `stop`) immediately after this log message is printed, causing intermittent failures if the `stop` call arrived before the `run` call finished. The change ensures that `stop` always occurs after `run`, avoiding intermittent failures in this test.

### Your checklist for this pull request

- [x] I have reviewed the [guidelines for contributing](../blob/master/CONTRIBUTING.md) to this repository.
- [x] I have added an entry to [History.md](../blob/master/History.md) if this PR fixes a bug or adds a feature. If it doesn't need an entry to HISTORY.md, I have added `[changelog skip]` or `[ci skip]` to the pull request title.
- [x] I have added appropriate tests if this PR fixes a bug or adds a feature.
- [x] My pull request is 100 lines added/removed or less so that it can be easily reviewed.
- [x] If this PR doesn't need tests (docs change), I added `[ci skip]` to the title of the PR.
- [x] If this closes any issues, I have added "Closes `#issue`" to the PR description or my commit messages.
- [x] I have updated the documentation accordingly.
- [x] All new and existing tests passed, including Rubocop.

### CODE REVIEW

File: test/test_puma_server.rb [hunk 1]
```diff
@@ -1141,4 +1141,12 @@ def test_client_quick_close_no_lowlevel_error_handler_call
     sleep 0.5
     assert_empty @events.stdout.string
   end
+
+  def test_run_stop_thread_safety
+    100.times do
```
Comment (nateberkopec) [inline test/test_puma_server.rb:1146]: How long does this take to run?

File: lib/puma/server.rb [hunk 2]
```diff
@@ -640,19 +641,16 @@ def graceful_shutdown
     end
 
     def notify_safely(message)
-      @check, @notify = Puma::Util.pipe unless @notify
-      begin
-        @notify << message
-      rescue IOError, NoMethodError, Errno::EPIPE
-         # The server, in another thread, is shutting down
+      @notify << message
```
Comment (nateberkopec) [inline lib/puma/server.rb:644]: Nice cleanup here.

File: test/test_puma_server.rb (hunk 1 above)
Comment (MSP-Greg) [reply test/test_puma_server.rb:1146]: I believe 0.3 sec on JRuby, less on MRI.  I should have reviewed this earlier, but...

I suspect I can revert the band-aids for JRuby (the sleep after run).  This is the same issue of `Server#run` getting messed up.

Comment (wjordan) [reply test/test_puma_server.rb:1146]: [0.02s](https://github.com/puma/puma/runs/1276840378#step:8:356) on MRI, [0.28s](https://github.com/puma/puma/runs/1276840522#step:9:422) on JRuby

**Review by MSP-Greg — APPROVED** [review]

**Pull request merged**